File: core/gateway/gpu_metrics_server.py
"""
GPU Metrics Server - Lightweight proxy for real GPU stats
Runs on host (not Docker) to access nvidia-smi directly
"""
import subprocess
import json
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse
import time

PORT = 8083

class GPUMetricsHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        path = urlparse(self.path).path
        
        if path == '/gpu-metrics' or path == '/':
            self.send_gpu_metrics()
        elif path == '/health':
            self.send_health()
        else:
            self.send_error(404, 'Not found')
    
    def send_gpu_metrics(self):
        try:
            result = subprocess.run(
                ["nvidia-smi", "--query-gpu=index,name,utilization.gpu,memory.used,memory.total,temperature.gpu",
                 "--format=csv,noheader,nounits"],
                capture_output=True,
                text=True,
                timeout=5
            )
            
            if result.returncode != 0:
                raise Exception(f"nvidia-smi failed: {result.stderr}")
            
            gpus = []
            for line in result.stdout.strip().split('\n'):
                parts = [p.strip() for p in line.split(',')]
                if len(parts) >= 6:
                    gpus.append({
                        "id": int(parts[0]),
                        "name": parts[1],
                        "utilization": float(parts[2]) if parts[2] else 0,
                        "memory_used_mb": float(parts[3]) if parts[3] else 0,
                        "memory_total_mb": float(parts[4]) if parts[4] else 1,
                        "temperature": float(parts[5]) if parts[5] else 0
                    })
            
            response = {
                "timestamp": time.time(),
                "gpus": gpus
            }
            
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps(response).encode())
            
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({"error": str(e)}).encode())
    
    def send_health(self):
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps({"status": "ok"}).encode())
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
    
    def log_message(self, format, *args):
        print(f"[GPU Metrics] {args[0]}")

if __name__ == "__main__":
    server = HTTPServer(('0.0.0.0', PORT), GPUMetricsHandler)
    print(f"GPU Metrics Server running on http://0.0.0.0:{PORT}")
    print(f"Endpoints: /gpu-metrics, /health")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down...")
        server.shutdown()

File: core/gateway/gpu_metrics_server.js
/**
 * GPU Metrics Server - Host-only, no Python needed.
 * Runs nvidia-smi and exposes JSON on port 8083.
 * Use: node gpu_metrics_server.js
 */
const http = require('http');
const { execFile } = require('child_process');

const PORT = 8083;
const CORS = { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' };

// CSV columns: index, name, utilization.gpu, memory.used, memory.total, temperature.gpu
// GPU name can contain commas (e.g. "NVIDIA RTX 4090, 24 GB"), so take last 4 cols as numbers.
function parseNvidiaSmi(stdout) {
  const gpus = [];
  const lines = stdout.trim().split('\n').filter(Boolean);
  for (let i = 0; i < lines.length; i++) {
    const parts = lines[i].split(',').map((p) => p.trim());
    if (parts.length >= 6) {
      const last4 = parts.slice(-4);
      const util = parseFloat(last4[0]) || 0;
      const usedMb = parseFloat(last4[1]) || 0;
      const totalMb = parseFloat(last4[2]) || 1;
      const temp = parseFloat(last4[3]) || 0;
      const name = parts.slice(1, -4).join(', ').trim() || parts[1] || 'GPU';
      gpus.push({
        id: parseInt(parts[0], 10) || i,
        name,
        utilization: util,
        memory_used_mb: usedMb,
        memory_total_mb: totalMb,
        temperature: temp,
      });
    }
  }
  return gpus;
}

function runNvidiaSmi() {
  return new Promise((resolve, reject) => {
    execFile(
      'nvidia-smi',
      [
        '--query-gpu=index,name,utilization.gpu,memory.used,memory.total,temperature.gpu',
        '--format=csv,noheader,nounits',
      ],
      { timeout: 5000, maxBuffer: 1024 * 1024 },
      (err, stdout, stderr) => {
        if (err) {
          reject(err.message || stderr || 'nvidia-smi failed');
          return;
        }
        try {
          resolve(parseNvidiaSmi(stdout));
        } catch (e) {
          reject(e.message);
        }
      }
    );
  });
}

const server = http.createServer(async (req, res) => {
  const path = (req.url || '/').split('?')[0];
  if (req.method === 'OPTIONS') {
    res.writeHead(200, { ...CORS, 'Access-Control-Allow-Methods': 'GET, OPTIONS', 'Access-Control-Allow-Headers': 'Content-Type' });
    res.end();
    return;
  }
  if (path === '/health') {
    res.writeHead(200, CORS);
    res.end(JSON.stringify({ status: 'ok' }));
    return;
  }
  if (path !== '/gpu-metrics' && path !== '/') {
    res.writeHead(404, CORS);
    res.end(JSON.stringify({ error: 'Not found' }));
    return;
  }
  try {
    const gpus = await runNvidiaSmi();
    res.writeHead(200, CORS);
    res.end(JSON.stringify({ timestamp: Date.now() / 1000, gpus }));
  } catch (e) {
    res.writeHead(500, CORS);
    res.end(JSON.stringify({ error: String(e.message || e) }));
  }
});

server.listen(PORT, '0.0.0.0', () => {
  console.log(`GPU Metrics Server running on http://0.0.0.0:${PORT}`);
  console.log('Endpoints: /gpu-metrics, /health');
});
